src/System.Composition.Hosting/src/System/Composition/Hosting/Core/CompositionOperation.cs

```diff
@@ -0,0 +1,121 @@
+    public sealed class CompositionOperation : IDisposable
+    {
+        private List<Action> _nonPrerequisiteActions;
+        private List<Action> _postCompositionActions;
+        private object _sharingLock;
+
+        // Construct using Run() method.
+        private CompositionOperation() { }
+
+        /// <summary>
+        /// Execute a new composition operation starting within the specified lifetime
+        /// context, for the specified activator.
+        /// </summary>
+        /// <param name="outermostLifetimeContext">Context in which to begin the operation (the operation can flow
+        /// to the parents of the context if requried).</param>
+        /// <param name="compositionRootActivator">Activator that will drive the operation.</param>
+        /// <returns>The composed object graph.</returns>
+        public static object Run(LifetimeContext outermostLifetimeContext, CompositeActivator compositionRootActivator)
+        {
+            Requires.ArgumentNotNull(outermostLifetimeContext, "outermostLifetimeContext");
+            Requires.ArgumentNotNull(compositionRootActivator, "compositionRootActivator");
+
+            using (var operation = new CompositionOperation())
+            {
+                var result = compositionRootActivator(outermostLifetimeContext, operation);
+                operation.Complete();
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// Called during the activation process to specify an action that can run after all
+        /// prerequesite part dependencies have been satisfied.
+        /// </summary>
+        /// <param name="action">Action to run.</param>
+        public void AddNonPrerequisiteAction(Action action)
+        {
+            if (action == null)
+                throw new ArgumentNullException("action");
+
+            if (_nonPrerequisiteActions == null)
+                _nonPrerequisiteActions = new List<Action>();
+
+            _nonPrerequisiteActions.Add(action);
+        }
+
+        /// <summary>
+        /// Called during the activation process to specify an action that must run only after
+        /// all composition has completed. See OnImportsSatisfiedAttribute.
+        /// </summary>
+        /// <param name="action">Action to run.</param>
+        public void AddPostCompositionAction(Action action)
+        {
+            Requires.ArgumentNotNull(action, "action");
+
+            if (_postCompositionActions == null)
+                _postCompositionActions = new List<Action>();
+
+            _postCompositionActions.Add(action);
+        }
+
+        internal void EnterSharingLock(object sharingLock)
+        {
+            Assumes.NotNull(sharingLock, "Sharing lock is required");
+
+            if (_sharingLock == null)
+            {
+                _sharingLock = sharingLock;
+                Monitor.Enter(sharingLock);
```

Comment:
Hey Stephen - here's the essential background, let me know if I can add more info. Will happily PR some documentation/comments that might be a bit lacking in places :-)

The core model for composition in this lib is a tree of `LifetimeContext` scopes, with longer lived scopes towards the root of the tree and short-lived ones towards the leaves. E.g. when representing components in a web application, the ones created in the root scope might be "singleton" and stick around for the life of the app, each concurrent web request might have a scope at the next level down for holding things like ORM sessions, and finally there might be some further nested scopes for managing the lifetimes of components used just for short transactions.

Composition operations build a graph of objects associated with a single `LifetimeContext` (usually starting at a leaf one). Sometimes these trigger more composition operations at other scopes up towards the root when instances of longer-lived components are required.

The scopes are shared at each level by the composition operations running on them, which is often done without contention, but when the operation hits a part whose instances are "shared" in that particular scope (e.g. all instances associated with a web request share the same ORM session) the sharing lock for that level is taken so that only one unique instance is created.

Rather than immediately releasing the lock, it has to be held for the duration of the composition operation's work in that scope, because some actions (e.g. property initialization) are deferred until other parts of the graph are created. The lock prevents shared parts from becoming visible to other composition operations until they're definitely fully constructed.

Hope this covers your question. All the best!